Add Archival Information Collections

Hwesta opened the pull request:
AICs are a special type of AIP that only contains a METS that contains information on each of the AIPs that are part of the AIC. 

An AIC is created by selecting some AIPs from the Archival Storage tab, entering the Dublin Core metadata, and a SIP is created.  Several new MicroServices added, notably createAICMETS.  

Supported by storage service changes artefactual/archivematica-storage-service#2

Hwesta committed:
Dashboard-side changes to start AIC creation chain

refs #5803

Added a CreateAICForm and submit button to the AIP search page that is
populated with the results information.  The results information is posted
to create_aic which creates an entry in a WatchedDirectory to start a
MicroServiceChain.  The directory is populated with files corresponding to
the AIPs that should go in the AIC - the filenames are the UUID, and the
contents are the AIP name.

Also, updated urls.py files to use url() instead of just tuples and did
minor cleanup.

Hwesta committed:
Dashboard-side changes to start AIC creation chain

refs #5803

Added a CreateAICForm and submit button to the AIP search page that is
populated with the results information.  The results information is posted
to create_aic which creates an entry in a WatchedDirectory to start a
MicroServiceChain.  The directory is populated with files corresponding to
the AIPs that should go in the AIC - the filenames are the UUID, and the
contents are the AIP name.

Also, updated urls.py files to use url() instead of just tuples and did
minor cleanup.

Hwesta committed:
Added createAICMETS client script

refs #5813

Added a client script to create the METS file for an AIC.  It collects AIP
information from files with an AIP uuid and contents of the AIP name, and
gets the AIP label by extracting the Dublin Core title from the AIP's METS
file.  Using that an the AIC's Dublin Core metadata, it creates the METS
file.

Updated archivematicaCommon/storageService.py to not rely on Django's ORM,
so all the client scripts that use it don't need to include Django's
settings file.  Added an extract file function to storageService that
fetches and saves a file from a Package (ie AIP)

Hwesta committed:
Add MicroServiceChainLinks for AIC creation

refs #5823

Added preliminary MicroServiceChainLinks for AIC creation, and other
necessary client scripts.  Currently does not allow for metadata entry.
Updated createAICMETS for new paths.

Hwesta committed:
Added metadata entry for AICs

refs #5823

Create AIC redirects to metadata entry page.  Added new form for Dublin
Core metadata for AIC that doesn't have Part of AIC, potential for more
divergent behaviour.  MicroServiceChainLink is started once metadata is
saved.  Cleaned up Dublin Core forms to be a ModelForm.

Hwesta committed:
ElasticSearch doesn't index processingMCP.xml

refs #5803

Put a filter in ElasticSearch functions to prevent processingMCP.xml from
being indexed.  Other paths use the METS file to find the files to index,
but the AIC METS file is in metadata/ not /.

Removed unused code, cleaned up imports, fixed removing tmp directories on
DB recreate.

Hwesta committed:
Fixed problems with AIC MicroServiceChainLinks

refs #5823

Added create METS.xml script to AIC chain.  Fixed createMETS file to not
bail if objects/metadata/transfers does not exist.

Updated bagit script to check that all the payload files/folders existed.
Added metadata/ folder to the bagit payload so the AIC METS file is
included.  Cleaned up bagit script.

Hwesta committed:
Added isPartOf to METS file

refs #5803

Renamed Part of AIP to be isPartOf, like Dublin Core name.  Updated Dublin
Core metadata section in create METS.  ElasticSearch should index all
the fields in the METS file, including the new isPartOf.

Hwesta committed:
Updated pointer file, storeAIP, and metadata entry for AIC

refs #5803

SIPs now store a sipType, which is "SIP" by default but is "AIC" for AICs.
AIP Pointer File stores the correct package type, pulled from the Dublin
Core metadata.  Store AIP sends package type of AIC when sipType is AIC.
Removed dependency on Django for store AIP.  The metadata entry form shows
the AIC form if the sipType is AIC, and the default DublinCore form
otherwise.

Also, fixed bug in dublin core METS creation.

Hwesta committed:
AIC METS file named with file UUID

refs #5803

Both the 'normal' METS file and the AIC-specific METS file were named the
same - METS.<AIP UUID>.xml.  The AIC-specific METS file in the metadata/
folder is the contents of the AIP, so it is now named with the file UUID
instead of the AIP UUID.

Hwesta committed:
Fixed METS file for AICs

refs #5883

Added metadata/ to the createMETS2 script, and added a check if the
directory exists before adding it to the structmap.  Also created a new
StandardTasksConfig for creating METS for an AIC that doesn't include the
amdSec.  Added the AIC METS file to Derivations, so it would work with
createMETS2.

Converted createAICMETS and createPointerFile to use
archivematicaXMLNamesSpace.  Renamed dctermsNS to dcNS, and removed unused
dcNS.

Hwesta committed:
Enable searching of AIPs by isPartOf

refs #5898

Added isPartOf and AICID to fields specially indexed for 'aips' and
'aipfiles'.  Parse the data to populate those fields out of the METS file.
Added options to drop-down menu to search by those fields.

Converted some functions to use namespaces in xml searches.

Hwesta committed:
AIC identifiers and isPartOf have AIC# prepended

refs #5803

Updated the metadata forms for AICs and AIPs to prepend 'AIC# ' where
needed.  Since DublinCore metadata in createAICMETS uses the createMETS2
function, removed get_aic_info from createAICMETS.  Updated display text
for searching on isPartOf.

Hwesta committed:
Index AIC METS file

refs #5898

When indexing an AIP/AIC, index files fileGrp USE='metadata' in addition to
'original'.  Since metadata files never have an amdSec, parse the UUID from
the file ID.

Hwesta committed:
Updated comment on how ElasticSearch searches

Hwesta committed:
Code cleanup and minor bug fixes

refs #5803

Fixed bugs if no DublinCore metadata entered.  Merged 'original' and
'metadata' file indexing.  Removed space after AIC# in isPartOf and
identifier.

Hwesta committed:
New show AIC checkbox and AIC identifier column in Archival Storage

refs #5898

Added 'Show AIC?' checkbox to Archival Storage search that displays a
column with the AIC identitfier (if AIC) or isPartOf field (if AIP).
Updated query parameter for file mode. Removed defunct StorageSearchForm.

Hwesta committed:
Added %SIPType% to SIP unit

SIP type was added to the database, but not available as a replacement
value.  Added it to the unitSIP replacementDic, and updated script that
used it.

Hwesta committed:
Add count of AIPs in AIC to ES and Archival Storage display

refs #5898

Want to be able to display the number of AIPs semantically included in an
AIC.  To do this, we index the number of AIPs in the AIC METS file into
ElasticSearch, and store that in the AIC METS file.

When creating the AIC METS file, we save the number of AIPS in a
UnitVariable for that AIC, and in the extent Dublin Core tag.  IndexAIP
looks for this variable if the SIP type is AIC and optionally passes it to
ElasticSearch, where it is retrieved by the Archival Storage tab.

Hwesta committed:
Fix download of AIC METS files

Hwesta committed:
Refactoring AIP reindex script

Hwesta committed:
Reindex AICs properly

refs #6034, #5803

Updated elasticsearch rebuild index from AIPs to check if it is an AIC, and
if so parse the number of AIPs out of the AIC METS file.

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaBagWithEmptyDirectories.py:
```diff
@@ -20,55 +20,79 @@
+    """ Run Bagit's create bag command. """
+    command = "/usr/share/bagit/bin/bag {}".format(arguments)
+    print 'Command to run:', command
+    exit_code, std_out, std_err = executeOrRun("command", command, printing=False)
```
I'd suggest either running this directly via a `subprocess` function, or going directly to `launchSubProcess`. Passing the command as a string leaves us open to potential shell-splitting issues, but since we're already [constructing](https://github.com/artefactual/archivematica/blob/dev/issue-5803-aic/src/MCPClient/lib/clientScripts/archivematicaBagWithEmptyDirectories.py#L81) the args as an array we can pass that directly to the commands being passed. That bypasses the shell altogether and directly execs the arguments.

`launchSubProcess` can take an array as the first argument, and doesn't require a string like `executeOrRun` does.

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaBagWithEmptyDirectories.py:
```diff
@@ -20,55 +20,79 @@
+    dir_list = get_sip_directories(sip_dir)
+
+    # Ensure all payload items actually exist
+    payload_entries = [e for e in args.payload_entries if os.path.exists(e)]
```
Maybe fold this into one iteration:

``` python
payload_entries = ['"{}"'.format(e) for e in args.payload_entries if os.path.exists(e)]
payload_entries = ' '.join(payload_entries)
```

Hwesta replied:
Put all the bagit args together in an array and passed that to executeOrRun instead.  All the string munging was to ensure that we passed a space-separated list of parameters as the command.  I suspect this used to directly run bagit, then needed some extra processing and this script was trying to emulate the old behavior.

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -561,6 +555,8 @@ def getIncludedStructMap():
     ret = []
     transferMetadata = os.path.join(baseDirectoryPath, "objects/metadata/transfers")
+    if not os.path.isdir(transferMetadata):
+        return []
```
:thumbsup:

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaXMLNamesSpace.py:
```diff
@@ -25,13 +25,11 @@
 metsNS = "http://www.loc.gov/METS/"
 premisNS = "info:lc/xmlns/premis-v2"
-dctermsNS = "http://purl.org/dc/terms/"
+dcNS = "http://purl.org/dc/terms/"
```
Were we not using the DC elements namespace? terms and elements are two different namespaces.

Hwesta replied:
I updated this before I knew that they were both useful namespaces.  dctermsNS was never actually used, so I removed it.

Probably as a separate commit, but should we move this to archivematicaCommon, and clean it up a little.

mistydemeo replied:
Yeah, I'd really like to do that. My METS code is going to have to carry around its own copy, but there are other places where this is useful outside createMETS.

mistydemeo commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -351,49 +364,95 @@ def index_mets_file_metadata(conn, uuid, metsFilePath, index, type, sipName, bac
-                xml = ElementTree.tostring(item)
+    # establish structure to be indexed for each file item
+    fileData = {
+        'archivematicaVersion': '1.0',
```
Maybe we should define a constant for this somewhere.

Hwesta replied:
Fixed in 501ff46ffe922b1aeef5026503266fcf3d4b5693, but I made these changes before that, and it got lost when I rebased.  Thanks.

mistydemeo commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -351,49 +364,95 @@ def index_mets_file_metadata(conn, uuid, metsFilePath, index, type, sipName, bac
+        filePath = file_.find('m:FLocat', namespaces=nsmap).attrib['{http://www.w3.org/1999/xlink}href']
+        indexData['filePath'] = filePath
+        _, fileExtension = os.path.splitext(filePath)
+        if fileExtension != '':
```
`if fileExtension:`?

mistydemeo commented on src/dashboard/src/components/archival_storage/views.py:
```diff
@@ -196,6 +220,45 @@ def search_augment_file_results(raw_results):
+def create_aic(request, *args, **kwargs):
+    aic_form = forms.CreateAICForm(request.POST or None)
+    if aic_form.is_valid():
+        results = ast.literal_eval(aic_form.cleaned_data['results'])
```
Where is this data coming from?

Hwesta replied:
From [elasticsearch](https://github.com/artefactual/archivematica/blob/dev/issue-5803-aic/src/dashboard/src/components/archival_storage/views.py#L150), it's used to get the info about the AIPs in the search (the AIPs that are supposed to go into the AIC) from ES to the frontend, to be sent back to Django to create the AIC.

mistydemeo commented on src/dashboard/src/components/ingest/forms.py:
```diff
@@ -16,25 +16,42 @@
+class DublinCoreMetadataForm(forms.ModelForm):
+    class Meta:
+        model = models.DublinCore
+        fields = ('title', 'is_part_of', 'creator', 'subject', 'description', 'publisher', 'contributor', 'date', 'format', 'identifier', 'source', 'relation', 'language', 'coverage', 'rights')
```
Wonder if it's worth this tuple in a centralized place somewhere - we use it in several places now.

mistydemeo replied:
Tuck it away as an attribute of the `models.DublinCore` class, maybe?

Hwesta committed:
Code review fixes

Bagit script updated to pass arguments as an array, not a space-separated
string.  Reverted changes to dcNS and dctermsNS.  Added dynamic
Archivematica version back in.

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaBagWithEmptyDirectories.py:
```diff
@@ -20,55 +20,80 @@
+
+    # Reconstruct bagit arguments
+    # Goal: bagit <operation> <destination> <flattened payload list> <optional args>
+    bagit_args = [args.operation, args.destination]
```
:thumbsup:

Hwesta committed:
Use existing temp directory

Instead of requiring new staging directory, use existing tmp directory.

mistydemeo commented on localDevSetup/recreateDB.sh:
```diff
@@ -28,6 +28,7 @@ password="demo"
 
 echo "Removing existing units"
 sudo ./removeUnitsFromWatchedDirectories.py
+sudo rm -rf /var/archivematica/sharedDirectory/tmp/tmp*
```
:thumbsup:
